### DISCUSSION TITLE
call actor from thread

### DISCUSSION THREAD

**Discussion by janjaapbos** [uazu/stakker#23] [Q&A] [answered]
When using a single Stakker in the main thread where the actor is also initialized, is it possible to call the actor from a sub thread?

If yes, I would really appreciate an example!

**Comment by janjaapbos** [comment]
Looking at #16 I think it needs to use a channel with the main thread to call the actor

**Comment by uazu** [comment, marked as answer]
As you have already figured out, Stakker itself doesn't provide a means to do this. Stakker intentionally doesn't include the event loop or any polling for external events (including events from other threads), to allow integration into any event loop or event system. That has to be layered below Stakker, e.g. using `stakker_mio` or whatever. So Stakker itself only manages the execution of calls within the same thread. A call from another thread would probably have to come through a channel (or similar), and notified to the event loop of the Stakker thread via a Waker, which tells the poll implementation (e.g. `stakker_mio`) to wake the main thread and call the handler, which reads the channel and executes the call.

It might be possible to simplify this down and provide the code built into Stakker if it's possible to find a simple common implementation which suits a lot of different scenarios. So I'm definitely open to adding this to Stakker at some point (like PipedThread) once the research/design/reduction/simplification work is done.

**Reply by janjaapbos** [reply]
Thank you very much. I have started to play with waker callbacks, passed in threads with an Arc Mutex.

**Reply by uazu** [reply]
I've added a new issue #24 to gather thoughts on this. This will likely get implemented sooner if I need it for work, simply due to time constraints. I have implemented something like a `RemoteFwd` previously, but I didn't think to generalize it at the time. However that didn't need cancellation (i.e. notification of destination gone away), which would be good to include in a general solution.
